### PR TITLE
Split itinerary and participant updates

### PR DESCRIPTION
**Pull request by gbiggs**
## New feature implementation

Closes open-rmf/rmf#8.

### Implemented feature

Reworks synchronisation between a database and mirrors so that participant information is sent separately from itinerary/route information. This enables scalable synchronisation that is robust to the timing of a mirror joining a query.

This PR is part of a set of three, all of which are needed for this feature.

- This PR, which implements the changes in the ROS 2 communication.
- open-rmf/rmf_traffic#14, which implements the changes in the core.
- open-rmf/rmf_internal_msgs#6, which implements the new messages required to support the feature.

This PR depends on the PRs that change mirror synchronisation to use a topic, and the PRs that enable updating participant footprints.

### Implementation description

A set of new messages have been created for sending participant descriptions via a topic.

The schedule node publishes a topic, `/rmf_traffic/participants`, which broadcasts all participant information when certain events occur:

- When a new participant registers
- When a participant unregisters
- When a mirror joins a query
- When a manual "send info" request is made

When a mirror receives the set of participants, it refreshes its own list to match.

To account for timing differences in receiving the list of participants and receiving the list of routes, which could mean that routes are received for participants not yet known about, a semi-caching mechanism is added. If a route is received that does not yet have a known participant, it is stored with an empty participant description. When views and snapshots are taken of a timeline, any participants/routes with empty descriptions are ignored (i.e. not given to the view).

When a mirror receives a set of participants, if it finds that a participant already exists it updates the description of that participant to match the newly-received description. A double pointer is used to store descriptions, ensuring that all places referencing the participant's description will receive the updated description immediately. To avoid snapshots having their descriptions changed (snapshots are meant to be immutable), they make a fresh copy of all descriptions when the snapshot is made.

### How to test

Launch the office demo world and start a couple of tasks. Rviz should show the visualisation of the scheduled routes. Kill the RMF rviz2 node that is publishing this visualisation; the visualisation will disappear from rviz. Start the RMF rviz2 node again manually, and the schedule visualisation should appear shortly.

### Notes for the reviewer

There is a lot of participant updating happening in the `rmf_traffic_schedule` node. If comments are added to the `Mirror::update_participants_info()` function, then this sort of output will be seen several times per second:

```
[rmf_traffic_schedule-1] [Mirror::update_participants_info] **********************
[rmf_traffic_schedule-1] [Mirror::update_participants_info] Replacing participant description 0
[rmf_traffic_schedule-1] [Mirror::update_participants_info] Replacing participant description 1
[rmf_traffic_schedule-1] [Mirror::update_participants_info] ======================
[rmf_traffic_schedule-1] [Mirror::update_participants_info] **********************
[rmf_traffic_schedule-1] [Mirror::update_participants_info] Replacing participant description 0
[rmf_traffic_schedule-1] [Mirror::update_participants_info] Replacing participant description 1
[rmf_traffic_schedule-1] [Mirror::update_participants_info] ======================
[rmf_traffic_schedule-1] [Mirror::update_participants_info] **********************
[rmf_traffic_schedule-1] [Mirror::update_participants_info] Replacing participant description 0
[rmf_traffic_schedule-1] [Mirror::update_participants_info] Replacing participant description 1
[rmf_traffic_schedule-1] [Mirror::update_participants_info] ======================
[rmf_traffic_schedule-1] [Mirror::update_participants_info] **********************
[rmf_traffic_schedule-1] [Mirror::update_participants_info] Replacing participant description 0
[rmf_traffic_schedule-1] [Mirror::update_participants_info] Replacing participant description 1
[rmf_traffic_schedule-1] [Mirror::update_participants_info] ======================
```

I am not sure why this is happening and if it should be considered a problem or not.

### CODE REVIEW

**Review by mxgrey — CHANGES_REQUESTED** [review]
This is looking really good. I just have a few relatively minor nitpicks/suggestions/observations.

File: rmf_traffic_ros2/include/rmf_traffic_ros2/StandardNames.hpp [hunk 1]
```diff
@@ -33,8 +33,9 @@ const std::string UnregisterParticipantSrvName = Prefix +
   "unregister_participant";
 const std::string RegisterQueryServiceName = Prefix + "register_query";
 const std::string UnregisterQueryServiceName = Prefix + "unregister_query";
-const std::string MirrorUpdateServiceName = Prefix + "mirror_update";
-const std::string MirrorWakeupTopicName = Prefix + "mirror_wakeup";
+const std::string ParticipantsInfoTopicName = Prefix + "participants";
+const std::string QueryUpdateTopicNameBase = Prefix + "query_update_";
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/include/rmf_traffic_ros2/StandardNames.hpp:37]: The underscore is intentional. This gets the ID of the query appended, giving topic names such as `query_update_1`.

File: rmf_traffic_ros2/include/rmf_traffic_ros2/schedule/MirrorManager.hpp [hunk 2]
```diff
@@ -76,13 +76,9 @@ class MirrorManager
...
   /// Attempt to update this mirror immediately.
   ///
-  /// \param[in] wait
-  ///   How long to block the current thread while waiting for the mirror to
-  ///   update. By default this will not block at all.
-  ///
   // TODO(MXG): Consider allowing this function to accept a callback that will
   // get triggered when the update is complete.
-  void update(rmf_traffic::Duration wait = rmf_traffic::Duration(0));
+  void update();
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/include/rmf_traffic_ros2/schedule/MirrorManager.hpp:81]: The catch is that there is no longer a wait. We could leave the parameter in and just ignore it, which would change the behaviour of the function with respect to that parameter but not break ABI.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/convert_ParticipantDescription.cpp [hunk 3]
```diff
@@ -18,6 +18,8 @@
 #include <rmf_traffic_ros2/schedule/ParticipantDescription.hpp>
 #include <rmf_traffic_ros2/Profile.hpp>
 
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/convert_ParticipantDescription.cpp]: Fixed in 24bdf40.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/Node.cpp [hunk 4]
```diff
@@ -340,32 +347,108 @@ void ScheduleNode::register_query(
...
+      update_topic_publisher,
+      std::nullopt);
+    mirror_update_topics.insert(std::make_pair(query_id, update_topic));
 
+    RCLCPP_INFO(
+      get_logger(),
+      "[" + std::to_string(query_id) + "] Registered query");
+  }
+  // Make sure the query watcher learns about the participants
+  broadcast_participants();
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/Node.cpp:442]: Fixed in 635df3c.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/Node.cpp [hunk 5]
```diff
@@ -340,32 +347,108 @@ void ScheduleNode::register_query(
...
+  // topic already exists. Return the query ID to the client without creating a
+  // new topic or query.
   response->query_id = query_id;
-  RCLCPP_INFO(
-    get_logger(),
-    "[" + std::to_string(query_id) + "] Registered query");
+
+  // TODO(Geoff): Trigger an update here so that the newly-registered query
+  // mirror gets the database contents sent to it. Otherwise it will have to
+  // wait until the next change is made in the database.
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/Node.cpp:449]: Done in 76ae35b.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/Node.cpp [hunk 6]
```diff
@@ -607,11 +736,38 @@ void ScheduleNode::publish_inconsistencies(
...
+  msg.database_version = database->latest_version();
+
+  for (auto query_it: registered_queries) {
+    msg.query_id = query_it.first;
+    const auto query_topic = mirror_update_topics.find(query_it.first);
+    if (query_topic != mirror_update_topics.end())
+    {
+      msg.patch = rmf_traffic_ros2::convert(
+        database->changes(query_it.second, query_topic->second.second));
+      query_topic->second.first->publish(msg);
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/Node.cpp]: Fixed in 272ee7c.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp [hunk 7]
```diff
@@ -48,119 +56,119 @@ class MirrorManager::Implementation
...
+        "[rmf_traffic_ros2::MirrorManager] Failed to deserialize Patch "
+        "message: " + std::string(e.what()));
+      // Get a full update in case we're just missing some information
+      request_update();
+    }
+  }
 
-    if (wait > rmf_traffic::Duration(0))
-      future.wait_for(wait);
+  void request_update(uint64_t minimum_version=0)
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp]: Fixed in b2643b5.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/Node.cpp [hunk 8]
```diff
@@ -381,18 +464,42 @@ void ScheduleNode::unregister_query(
...
-  registered_queries.erase(it);
-  response->confirmation = true;
+  // Find the update publisher for this query
+  const auto query_topic = mirror_update_topics.find(request->query_id);
+  if (query_topic != mirror_update_topics.end())
+  {
+    // If the topic has just one subscriber (or zero, in case there is some kind
+    // of race condition with the subscriber) then remove the publisher and the
+    // query.
+    if (query_topic->second.first->get_subscription_count() <= 1)
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/Node.cpp]: a8c95de adds a counter to the query topics so they can be removed when the number of registered mirrors reaches zero.

File: rmf_traffic_ros2/include/rmf_traffic_ros2/schedule/MirrorManager.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mxgrey) [reply rmf_traffic_ros2/include/rmf_traffic_ros2/schedule/MirrorManager.hpp:81]: I think it's fine to break ABI here. I have very very strong doubts that anyone is using this function anyhow.

**Review by mxgrey — CHANGES_REQUESTED** [review]
Just one recommendation about the tuple.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/internal_Node.hpp [hunk 9]
```diff
@@ -118,20 +122,24 @@ class ScheduleNode : public rclcpp::Node
...
-
-  void mirror_update(
-    const request_id_ptr& request_header,
-    const MirrorUpdate::Request::SharedPtr& request,
-    const MirrorUpdate::Response::SharedPtr& response);
-
-  MirrorUpdateService::SharedPtr mirror_update_service;
+  using MirrorUpdate = rmf_traffic_msgs::msg::MirrorUpdate;
+  using MirrorUpdateTopicPublisher = rclcpp::Publisher<MirrorUpdate>::SharedPtr;
+  using MirrorUpdateTopic = std::tuple<
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/internal_Node.hpp]: I couldn't agree more about the ugliness of tuples. I changed it to a struct in 62e1193.

**Review by mxgrey — APPROVED** [review]
The changes seem to be working :+1:

**Pull request merged**